File: src/python/problems/find_special_integer_test.py
# https://leetcode.com/problems/element-appearing-more-than-25-in-sorted-array

from core.problem_base import *

class FindSpecialInteger(ProblemBase):
    def Solution1(self, arr: list[int]) -> int:
        minCount = len(arr) // 4
        for i in range(len(arr) - minCount):
            if arr[i] == arr[i + minCount]:
                return arr[i]

        return arr[0]
    
    def Solution2(self, arr: list[int]) -> int:
        minCount = len(arr) // 4
        count = 1
        for i in range(1, len(arr)):
            count = count + 1 if arr[i - 1] == arr[i] else 1
            if count > minCount:
                return arr[i]

        return arr[0]

if __name__ == '__main__':
    TestGen(FindSpecialInteger) \
        .Add(lambda tc: tc.Param([1,2,2,6,6,6,6,7,10]).Result(6)) \
        .Add(lambda tc: tc.Param([1,1]).Result(1)) \
        .Add(lambda tc: tc.Param([1,1,2,2,3,3,3,3]).Result(3)) \
        .Run()

File: src/go/problems/find_special_integer_test.go
// https://leetcode.com/problems/element-appearing-more-than-25-in-sorted-array

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type FindSpecialInteger struct{}

func TestFindSpecialInteger(t *testing.T) {
	gen := core.TestSuite[FindSpecialInteger]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{1,2,2,6,6,6,6,7,10}).Result(6)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1,1}).Result(1)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1,1,2,2,3,3,3,3}).Result(3)
	}).Run(t)
}

func (FindSpecialInteger) Solution1(arr []int) int {
	minCount := len(arr) / 4
	for i := 0; i < len(arr) - minCount; i++ {
		if arr[i] == arr[i + minCount] {
			return arr[i]
		}
	}

	return arr[0]
}

func (FindSpecialInteger) Solution2(arr []int) int {
	minCount := len(arr) / 4;
	count := 1
	for i := 1; i < len(arr); i++ {
		if arr[i - 1] == arr[i] {
			count++
		} else {
			count = 1
		}

		if count > minCount {
			return arr[i]
		}
	}

	return arr[0]
}

